File: data_check.py
#!/usr/bin/python
"""
With this module we get the POIs with the tags vegan = * and
vegetarian = * from OpenStreetMap and fill them in a file.
"""

import os         # for handling files
import time       # for sleep
import json       # read and write json
import sys        # to check the python version
import datetime   # for the timestamp
#import urllib.request
import urllib.parse

#import httplib2   # to check if a website is reachable

from email_validator import validate_email, EmailNotValidError

import urllib3    # for the HTTP GET request

assert sys.version_info >= (3, 0)

# constants for the overpass request

# # server list (from: https://wiki.openstreetmap.org/wiki/Overpass_API)
SERVERS = [
    "https://lz4.overpass-api.de/api/interpreter",
    "https://overpass.kumi.systems/api/interpreter",
    "https://z.overpass-api.de/api/interpreter",
    "http://api.openstreetmap.fr/api/interpreter",
    "http://dev.overpass-api.de/api_drolbr/interpreter",
    "http://overpass-api.de/api/interpreter",
    "http://overpass.openstreetmap.fr/api/interpreter"
]
HTTP = urllib3.PoolManager()

# # constants for the output files
TIMESTAMP = str(datetime.datetime.now())                             # the actual date and time
DATE = datetime.datetime.now().strftime("%Y-%m-%d")                  # the actual date
DATA_DIR = os.path.dirname(os.path.abspath(__file__))                # get the path of the directory of this script
VEGGIEPLACES_TEMPFILE = DATA_DIR + "/data/places_temp.json"          # the temp file to store the data
VEGGIEPLACES_TEMPFILE_MIN = DATA_DIR + "/data/places_temp.min.json"  # the minimized temp file
VEGGIEPLACES_TEMPFILE_GZIP = DATA_DIR + "/data/places_temp.min.json.gz"  # the gzipped temp file
VEGGIEPLACES_FILE = DATA_DIR + "/data/places.json"                   # the data file which will be used for the map
VEGGIEPLACES_FILE_MIN = DATA_DIR + "/data/places.min.json"           # the minimized data file which will be used for the map
VEGGIEPLACES_FILE_GZIP = DATA_DIR + "/data/places.min.json.gz"       # the gzipped data file which will be used for the map
VEGGIESTAT_FILE = DATA_DIR + "/data/stat.json"                       # the statistics data file which will be used for the map
VEGGIEPLACES_OLDFILE = DATA_DIR + "/data/places_old.json"            # previous version of the data file (helpful to examine changes)
OVERPASS_FILE = DATA_DIR + "/data/overpass.json"                     # the raw overpass output file (useful for later use)
VEGGIEPLACES_CHECKRESULT_FILE = DATA_DIR + "/data/check_results.json"   # check results


# variables to handle the json data
places_data = {}
places_data_checks = {}
stat_data = {}




def get_data_osm():
    """Get the data from OSM."""
    # Initialize variables
    server = 0
    result = None

    # Sending a request to one server after another until one gives a valid answer or
    # the end of the server list is reached.
    while (server < len(SERVERS)) and (result is None):


        #urx = "https://veggiekarte.de/data/overpass.json"

        urx = "http://localhost/veggiekarte/veggiekarte-local-testing/data/overpass.json"

        osm_request = HTTP.request("GET", urx)



        # Check the status of the request
        if osm_request.status == 200:
            print("Received answer successfully.")

            # Store the raw output in a file (for any later use)
            with open(OVERPASS_FILE, "wb") as overpass_file:
                overpass_file.write(osm_request.data)

            result = json.loads(osm_request.data.decode("utf-8"))
        elif osm_request.status == 400:
            print("HTTP error code ", osm_request.status, ": Bad Request")
            time.sleep(5)
        elif osm_request.status == 429:
            print("HTTP error code ", osm_request.status, ": Too Many Requests")
            time.sleep(60)
        elif osm_request.status == 504:
            print("HTTP error code ", osm_request.status, ": Gateway Timeout")
            time.sleep(600)
        else:
            print("Unknown HTTP error code: ", osm_request.status)

        # Increase to get another server for the next pass of the loop.
        server += 1

    return result




def uri_validator(uri, with_path):
    try:
        result = urllib.parse.urlparse(uri)
        if with_path:
            return all([result.scheme, result.netloc, result.path])
        else:
            return all([result.scheme, result.netloc])
    except:
        return False



def check_data(data):

    # Adding timestamp
    places_data["_timestamp"] = TIMESTAMP

    places_data["type"] = "FeatureCollection"

    # Adding list object which will contain all place objects
    places_data["features"] = []

    places_data_checks["_timestamp"] = TIMESTAMP
    places_data_checks["type"] = "FeatureCollection"
    places_data_checks["features"] = []

    # Variables to print progress in the console
    osm_element_index = 0
    osm_elements_number = len(data["elements"])

    # Go through every osm element and put the information into a new places element.
    for osm_element in data["elements"]:

        element_id = osm_element["id"]
        element_type = osm_element["type"]
        tags = osm_element.get("tags", {})

        place_check_obj = {}
        place_check_obj["type"] = "Feature"
        place_check_obj["properties"] = {}
        place_check_obj["properties"]["_id"] = element_id
        place_check_obj["properties"]["_type"] = element_type
        place_check_obj["properties"]["undefined"] = []
        place_check_obj["properties"]["issues"] = []

        if element_type == "node":
            lat = osm_element.get("lat", None)
            lon = osm_element.get("lon", None)
        elif element_type == "way" or element_type == "relation":
            center_coordinates = osm_element.get("center", None)  # get the coordinates from the center of the object
            lat = center_coordinates.get("lat", None)
            lon = center_coordinates.get("lon", None)

        place_check_obj["geometry"] = {}
        place_check_obj["geometry"]["type"] = "Point"
        place_check_obj["geometry"]["coordinates"] = [lon, lat]

        # Name
        if "name" in tags:
            name = tags["name"]
            place_check_obj["properties"]["name"] = name
        else:
            # If there is no name, take the english if exists
            if "name:en" in tags:
                name = tags["name:en"]
            # If it is a vending machine, name it "vending machine"
            elif "amenity" in tags:
                if tags["amenity"] == "vending_machine":
                  name = "vending machine"
            else:
                # If there is no name given from osm, we build one
                name = "%s %s" % (element_type, element_id)
                # Log this
                place_check_obj["properties"]["undefined"].append("name")
        # Double quotes could escape code, so we have to replace them:
        name = name.replace('"', '”')
        place_check_obj["properties"]["name"] = name

        osm_element_index += 1
        print(osm_element_index, ' / ', osm_elements_number, '\t', end='\r')



        # Diet tags
        if "diet:vegan" in tags:
            diet_vegan = tags.get("diet:vegan", "")
            if diet_vegan != "only" and diet_vegan != "yes" and diet_vegan != "limited" and diet_vegan != "no":
                place_check_obj["properties"]["issues"].append("'diet:vegan' has an unusual value: " + diet_vegan)
        if "diet:vegetarian" in tags:
            diet_vegetarian = tags.get("diet:vegetarian", "")
            if diet_vegetarian != "only" and diet_vegetarian != "yes" and diet_vegetarian != "no":
                place_check_obj["properties"]["issues"].append("'diet:vegetarian' has an unusual value: " + diet_vegetarian)

        # Cuisine
        #if "cuisine" not in tags:
        #    if "shop" not in tags:
        #        if tags.get("amenity", "") != "cafe":
        #            place_check_obj["properties"]["undefined"].append("cuisine")

        # Address
        if "addr:street" not in tags:
            place_check_obj["properties"]["undefined"].append("addr:street")
        if "addr:housenumber" not in tags:
            place_check_obj["properties"]["undefined"].append("addr:housenumber")
        if not "addr:city" in tags:
            if not "addr:suburb" in tags:
                place_check_obj["properties"]["undefined"].append("addr:city/suburb")
        if not "addr:postcode" in tags:
            place_check_obj["properties"]["undefined"].append("addr:postcode")

        # Website (till now we only check if the URI is valid, not if the website is reachable)
        website = 'undefined'
        if "contact:website" in tags:
            website = tags.get("contact:website", "")
            if uri_validator(website, False) is False:
                place_check_obj["properties"]["issues"].append("'contact:website' URI invalid")
        if "website" in tags:
            website = tags.get("website", "")
            if uri_validator(website, False) is False:
                place_check_obj["properties"]["issues"].append("'website' URI invalid")
        if "facebook" in website:
            place_check_obj["properties"]["issues"].append("'facebook' URI as website -> change to 'contact:facebook'")
        if "instagram" in website:
            place_check_obj["properties"]["issues"].append("'instagram' URI as website -> change to 'contact:instagram'")
        if "contact:website" in tags and "website" in tags:
            place_check_obj["properties"]["issues"].append("'website' and 'contact:website' defined -> remove one")

        # Facebook
        if "contact:facebook" in tags:
            contact_facebook = tags.get("contact:facebook", "")
            if contact_facebook.startswith("http://"):
                place_check_obj["properties"]["issues"].append("'contact:facebook' starts with 'http' instead of 'https'")
            elif not contact_facebook.startswith("https://www.facebook.com/"):
                place_check_obj["properties"]["issues"].append("'contact:facebook' does not starts with 'https://www.facebook.com/'")
            elif uri_validator(contact_facebook, True) is False:
                place_check_obj["properties"]["issues"].append("'contact:facebook' URI invalid")
        if "facebook" in tags:
            place_check_obj["properties"]["issues"].append("old tag: 'facebook' -> change to 'contact:facebook'")

        # Instagram
        if "contact:instagram" in tags:
            contact_instagram = tags.get("contact:instagram", "")
            if contact_instagram.startswith("http://"):
                place_check_obj["properties"]["issues"].append("'contact:instagram' starts with 'http' instead of 'https'")
            elif not contact_instagram.startswith("https://www.instagram.com/"):
                place_check_obj["properties"]["issues"].append("'contact:instagram' does not starts with 'https://www.instagram.com/'")
            elif uri_validator(contact_instagram, True) is False:
                place_check_obj["properties"]["issues"].append("'contact:instagram' URI invalid")
        if "instagram" in tags:
            place_check_obj["properties"]["issues"].append("old tag 'instagram'")

        # E-Mail
        if "contact:email" in tags:
            email = tags.get("contact:email", "")
        elif "email" in tags:
            email = tags.get("email", "")
        if "contact:email" in tags or "email" in tags:
            try:
              valid = validate_email(email)
            except EmailNotValidError as e:
              place_check_obj["properties"]["issues"].append("E-Mail is not valid: " + str(e))
        if "contact:email" in tags and "email" in tags:
            place_check_obj["properties"]["issues"].append("'email' and 'contact:email' defined -> remove one")

        # Phone
        if "contact:phone" in tags:
            contact_phone = tags.get("contact:phone", "")
            if not contact_phone.startswith("+"):
                place_check_obj["properties"]["issues"].append("'contact:phone' has no international format like '+44 20 84527891'")
        if "phone" in tags:
            phone = tags.get("phone", "")
            if not phone.startswith("+"):
                place_check_obj["properties"]["issues"].append("'phone' has no international format like '+44 20 84527891'")
        if "contact:phone" in tags and "phone" in tags:
            place_check_obj["properties"]["issues"].append("'phone' and 'contact:phone' defined -> remove one")

        # Opening hours
        opening_hours = 'undefined'
        if "opening_hours:covid19" in tags and tags["opening_hours:covid19"] != "same" and tags["opening_hours:covid19"] != "restricted":
            opening_hours = tags["opening_hours:covid19"]
        elif "opening_hours" in tags:
            opening_hours = tags["opening_hours"]
        else:
            place_check_obj["properties"]["undefined"].append("opening_hours")
        if "\n" in opening_hours or "\r" in opening_hours:
            place_check_obj["properties"]["issues"].append("There is a line break in 'opening_hours' -> remove")

        # Disused
        if "disused" in "".join(tags):
            place_check_obj["properties"]["issues"].append("There is a 'disused' tag: Check whether this tag is correct. If so, please remove the diet tags.")

        # Count issues
        place_check_obj["properties"]["issue_number"] = len(place_check_obj["properties"]["issues"]) + len(place_check_obj["properties"]["undefined"])
        
        if len(place_check_obj["properties"]["issues"]) == 0:
            del(place_check_obj["properties"]["issues"])
        if len(place_check_obj["properties"]["undefined"]) == 0:
            del(place_check_obj["properties"]["undefined"])

        # Only use elements with issues
        if place_check_obj["properties"]["issue_number"] > 0:
            places_data_checks["features"].append(place_check_obj)
    print(osm_elements_number, ' elements.')



def main():
    """Call the functions to get and write the osm data."""
    # Get data
    if len(sys.argv) < 2:
        osm_data = get_data_osm()
    else:
        # For testing without new osm requests
        osm_data = json.load(open(sys.argv[1]))

    # Write data
    if osm_data is not None:
        check_data(osm_data)

        # Write check result file in pretty format
        outfile = open(VEGGIEPLACES_CHECKRESULT_FILE, "w")
        outfile.write(json.dumps(places_data_checks, indent=1, sort_keys=True))
        outfile.close()

    else:
        print("A problem has occurred. The old VEGGIE_MAP was not replaced!")

main()

File: datacheck/datacheck.js
#!/usr/bin/env node
/* eslint-disable camelcase */
/* eslint-disable no-await-in-loop */

/* eslint-disable complexity */
/**
 * Node.js (ESM) port of the original Python datacheck script.
 * Checks OpenStreetMap place data and produces validation / issue output.
 * Original Python logic preserved as close as practical.
 */
import OpeningHours from "opening_hours";
import { fileURLToPath } from "node:url";
import fs from "node:fs/promises";
import { parsePhoneNumberFromString } from "libphonenumber-js";
import path from "node:path";
import process from "node:process";
import validator from "email-validator";

// ---------------- Constants ----------------
const TIMESTAMP = new Date().toISOString();
const DATE = TIMESTAMP.slice(0, 10); // YYYY-MM-DD
const __filename = fileURLToPath(import.meta.url);
const __dirname = path.dirname(__filename);
const DATA_PATH = path.resolve(__dirname, "..", "data");
const OVERPASS_FILE = path.join(DATA_PATH, "overpass.json");
const VEGGIEPLACES_CHECK_RESULT_FILE = path.join(DATA_PATH, "check_results.json");
const URL_DATA_FILE = path.join(DATA_PATH, "urldata.json");

const MAX_URL_CHECKS = 100; // Limit external requests
const MIN_URL_CHECK_AGE = 50; // Days before re-checking a URL
const HTTP_TIMEOUT_MS = 3000; // Shorter timeout for speed
const FETCH_CONCURRENCY = 25; // Higher concurrency for network phase
const SKIP_FULL_CHECK_DOMAINS = ["instagram.com", "www.instagram.com", "facebook.com", "www.facebook.com"]; // Heavy domains -> skip network

const REQUEST_HEADERS = {
  "User-Agent": "Mozilla/5.0 (X11; Linux x86_64) AppleWebKit/537.36 (KHTML, like Gecko) VeggieKarteDatacheck/1.0 Safari/537.36"
};

class Processor {
  counter = 0;
  urlData = {}; // Plain object cache
}
const proc = new Processor();

async function readJson(file) {
  try {
    const txt = await fs.readFile(file, "utf-8");
    return JSON.parse(txt);
  }
  catch (err) {
    if (err.code === "ENOENT") {
      return null;
    }
    throw err;
  }
}

async function writeJson(file, obj) {
  const txt = JSON.stringify(obj, null, 1);
  await fs.writeFile(file, txt, "utf-8");
}

function isUrlFormatValid(url) {
  try {
    const parsed = new URL(url);
    return Boolean(parsed.protocol && parsed.hostname);
  }
  catch {
    return false;
  }
}

async function fetchWithTimeout(url, opts = {}) {
  const controller = new AbortController();
  const timeout = setTimeout(() => controller.abort(), HTTP_TIMEOUT_MS);
  try {
    const response = await fetch(url, { ...opts, signal: controller.signal, headers: REQUEST_HEADERS });
    return response;
  }
  catch (err) {
    if (err.name === "AbortError") {
      return { status: 499, aborted: true };
    }
    // Convert all errors into synthetic failed response object to avoid throwing upstream
    return { status: 598, error: err };
  }
  finally {
    clearTimeout(timeout);
  }
}

async function isUrlOk(url) {
  const result = { date: DATE };
  if (Object.hasOwn(proc.urlData, url)) {
    const cachedResult = proc.urlData[url];
    result.isOk = cachedResult.isOk;
    result.text = cachedResult.text;
    return result;
  }
  // Quick skip for known heavy social media hosts
  try {
    const host = new URL(url).hostname.toLowerCase();
    if (SKIP_FULL_CHECK_DOMAINS.includes(host)) {
      result.isOk = true;
      result.text = "Skipped domain - assumed OK";
      proc.urlData[url] = result;
      return result;
    }
  }
  catch { /* Ignore parse errors here */ }
  if (!isUrlFormatValid(url)) {
    result.isOk = false;
    result.text = "No valid URL format";
  }
  else if (proc.counter >= MAX_URL_CHECKS) {
    result.isOk = true;
    result.text = "Not checked";
  }
  else {
    try {
      // Try lightweight HEAD first
      let response = await fetchWithTimeout(url, { method: "HEAD" });
      if (response.status === 405 || response.status === 501) {
        // Server does not allow HEAD -> retry with GET
        response = await fetchWithTimeout(url, { method: "GET" });
      }
      if (response.status < 400) {
        result.isOk = true;
        result.text = "OK";
      }
      else if (response.status === 403) {
        result.isOk = true;
        result.text = "Can't do full check: HTTP response: Forbidden";
      }
      else if (response.status === 429) {
        result.isOk = true;
        result.text = "Can't do full check: HTTP response: Too Many Requests";
      }
      else if (response.status === 499) {
        result.isOk = true; // Treat aborted (timeout) as soft-ok to not flood issues
        result.text = "Timeout (soft pass)";
      }
      else if (response.status === 598) {
        result.isOk = false;
        result.text = `Network error: ${response.error?.name || "Unknown"}`;
      }
      else {
        result.isOk = false;
        result.text = "HTTP response error";
        console.log(url, response.status);
      }
      result.text = `${result.text} - ${response.status}`;
      proc.counter += 1;
    }
    catch (err) {
      result.isOk = false;
      result.text = `Exception: ${err.name}`;
      console.log(url, result.text);
    }
  }
  proc.urlData[url] = result;
  return result;
}

// Run tasks with concurrency limit
async function runWithLimit(items, limit, worker) {
  const results = new Array(items.length);
  let index = 0;
  async function next() {
    const current = index;
    if (current >= items.length) { return; }
    index += 1;
    try {
      results[current] = await worker(items[current], current);
    }
    catch (err) {
      results[current] = { error: err };
    }
    next();
  }
  const starters = Array.from({ length: Math.min(limit, items.length) }, () => next());
  await Promise.all(starters);
  return results;
}

// Collect unique URLs needing checks (excluding already cached or skipped)
function collectUrls(elements) {
  const urls = new Set();
  for (const el of elements) {
    const tags = el.tags || {};
    ["contact:website", "website", "contact:facebook", "contact:instagram"].forEach((tag) => {
      if (tag in tags) {
        let tagValue = tags[tag];
        if (tag === "contact:facebook" && !tagValue.startsWith("http")) { tagValue = `https://www.facebook.com/${tagValue}`; }
        if (tag === "contact:instagram" && !tagValue.startsWith("http")) { tagValue = `https://www.instagram.com/${tagValue}`; }
        if (!Object.hasOwn(proc.urlData, tagValue)) {
          urls.add(tagValue);
        }
      }
    });
  }
  return Array.from(urls);
}

function normalizeName(name) {
  return name.replace(/"/gu, "”");
}
function validateEmailAddress(email) {
  return validator.validate(email);
}

function checkPhoneNumber(issueCollector, tagName, tags) {
  const raw = tags[tagName] || "";
  const invalidChars = raw.replace(/[0-9 +;-]/gu, "");
  if (invalidChars.length) {
    issueCollector.push(`'${tagName}' contains characters that are not allowed: '${invalidChars}'`);
  }
  const first = raw.split(";")[0];
  try {
    const phoneNumber = parsePhoneNumberFromString(first);
    if (!phoneNumber || !phoneNumber.isValid()) {
      issueCollector.push(`'${tagName}': Validation of number '${first}' failed. Is this number correct?.`);
      return;
    }
    const intl = phoneNumber.formatInternational();
    const e164 = phoneNumber.format("E.164");
    const rfc3966 = phoneNumber.format("RFC3966").replace("tel:", "");
    if (![intl, e164, rfc3966].includes(first)) {
      issueCollector.push(`'${tagName}' does not conform to the ITU-T E.123, E.164 or RFC 3966 pattern. It's '${first}' but '${intl}' (E.123) is recommended.`);
    }
  }
  catch (err) {
    issueCollector.push(`'${tagName}' not corresponds to the ITU-T E.123, E.164 nor RFC 3966 pattern - Error message: ${err.message}`);
  }
}

function checkOpeningHours(openingHoursValue, issues) {
  if (openingHoursValue.includes("\n") || openingHoursValue.includes("\r")) {
    issues.push("There is a line break in 'opening_hours' -> remove");
  }
  if (openingHoursValue.includes("SH")) {
    return;
  }
  try {
    const opening = new OpeningHours(openingHoursValue);
    const warnings = opening.getWarnings?.() || [];
    warnings.forEach(warning => issues.push(`opening_hours: ${warning}`));
  }
  catch (err) {
    issues.push(`opening_hours: ${err.message}`);
  }
}

function sortByIssueCount(features) {
  return features.sort((featureA, featureB) => featureB.properties.issue_count - featureA.properties.issue_count);
}
function containsDisused(tags) {
  return Object.entries(tags).some(([key, value]) => `${key}=${value}`.includes("disused"));
}

async function checkData(osmData) {
  const features = [];
  const elements = osmData?.elements || [];
  const total = elements.length;

  // Phase 1: batch URL checks
  const urlsToCheck = collectUrls(elements);
  const limitedList = urlsToCheck.slice(0, MAX_URL_CHECKS); // Enforce cap
  await runWithLimit(limitedList, FETCH_CONCURRENCY, url => isUrlOk(url));

  // Phase 2: element processing using cached results
  for (let index = 0; index < total; index += 1) {
    const osmElement = elements[index];
    const elementId = osmElement.id;
    const elementType = osmElement.type;
    const tags = osmElement.tags || {};
    process.stdout.write(`${index + 1} / ${total}\r`);
    let lat;
    let lon;
    if (elementType === "node") {
      lat = osmElement.lat;
      lon = osmElement.lon;
    }
    else if (elementType === "way" || elementType === "relation") {
      const center = osmElement.center || {};
      lat = center.lat;
      lon = center.lon;
    }
    const properties = { _id: elementId, _type: elementType, undefined: [], issues: [] };
    const geometry = { type: "Point", coordinates: [lon, lat] };
    let name = tags.name || tags["name:en"] || (tags.amenity === "vending_machine" ? "vending machine" : `${elementType} ${elementId}`);
    if (!tags.name) { properties.undefined.push("name"); }
    name = normalizeName(name);
    properties.name = name;
    if ("diet:vegan" in tags) {
      const dietVegan = tags["diet:vegan"];
      properties.diet_vegan = dietVegan;
      if (!["only", "yes", "limited", "no"].includes(dietVegan)) {
        properties.issues.push(`'diet:vegan' has an unusual value: ${dietVegan}`);
      }
      else if ("diet:vegetarian" in tags && dietVegan === "only") {
        const dietVegetarian = tags["diet:vegetarian"];
        if (dietVegetarian !== "only") { properties.issues.push("'diet:vegan' is 'only' then 'diet:vegetarian' is unneccessary or should be 'only' too."); }
      }
    }
    else {
      properties.undefined.push("diet:vegan");
    }
    if (tags["diet:vegan"] !== "no") {
      if (!("cuisine" in tags) && !("shop" in tags)) {
        if (!["cafe", "ice_cream", "bar"].includes(tags.amenity || "")) { properties.undefined.push("cuisine"); }
      }
      if ("cuisine" in tags) {
        const cuisine = tags.cuisine;
        if (cuisine.includes("vegan")) { properties.issues.push("There is 'vegan' in the cuisine tag. Remove it and create a 'diet:vegan' tag if there is none."); }
        if (cuisine.includes("vegetarian")) { properties.issues.push("There is 'vegetarian' in the cuisine tag. Remove it and create a 'diet:vegetarian' tag if there is none."); }
      }
      let website = "undefined";
      if ("contact:website" in tags) {
        website = tags["contact:website"];
        const check = await isUrlOk(website);
        if (!check.isOk) { properties.issues.push(`'contact:website' ${check.text}`); }
      }
      if ("website" in tags) {
        website = tags.website;
        const check = await isUrlOk(website);
        if (!check.isOk) { properties.issues.push(`'website' ${check.text}`); }
      }
      if (website.includes("facebook")) { properties.issues.push("'facebook' URI as website -> change to 'contact:facebook'"); }
      if (website.includes("instagram")) { properties.issues.push("'instagram' URI as website -> change to 'contact:instagram'"); }
      if ("contact:website" in tags && "website" in tags) { properties.issues.push("'website' and 'contact:website' defined -> remove one"); }
      if ("contact:facebook" in tags) {
        let contactFacebook = tags["contact:facebook"];
        if (contactFacebook.startsWith("http://")) {
          properties.issues.push("'contact:facebook' starts with 'http' instead of 'https'");
        }
        else if (contactFacebook.startsWith("https://")) {
          if (contactFacebook.startsWith("https://www.facebook.com/") || contactFacebook.startsWith("https://facebook.com/")) {
            const check = await isUrlOk(contactFacebook);
            if (!check.isOk) { properties.issues.push(`'contact:facebook' ${check.text}`); }
          }
          else {
            properties.issues.push("'contact:facebook' should start with 'https://www.facebook.com/'");
          }
        }
        else {
          contactFacebook = `https://www.facebook.com/${contactFacebook}`;
          const check = await isUrlOk(contactFacebook);
          if (!check.isOk) { properties.issues.push(`'contact:facebook' ${check.text}`); }
        }
      }
      if ("facebook" in tags) { properties.issues.push("old tag: 'facebook' -> change to 'contact:facebook'"); }
      if ("contact:instagram" in tags) {
        let contactInstagram = tags["contact:instagram"];
        if (contactInstagram.startsWith("http://")) {
          properties.issues.push("'contact:instagram' starts with 'http' instead of 'https'");
        }
        else if (contactInstagram.startsWith("https://")) {
          if (contactInstagram.startsWith("https://www.instagram.com/") || contactInstagram.startsWith("https://instagram.com/")) {
            const check = await isUrlOk(contactInstagram);
            if (!check.isOk) { properties.issues.push(`'contact:instagram' ${check.text}`); }
          }
          else {
            properties.issues.push("'contact:instagram' should start with 'https://www.instagram.com/'");
          }
        }
        else {
          contactInstagram = `https://www.instagram.com/${contactInstagram}`;
          const check = await isUrlOk(contactInstagram);
          if (!check.isOk) { properties.issues.push(`'contact:instagram' ${check.text}`); }
        }
      }
      if ("instagram" in tags) { properties.issues.push("old tag 'instagram' -> change to 'contact:instagram'"); }
      let email;
      if ("contact:email" in tags) { email = tags["contact:email"]; }
      else if ("email" in tags) { email = tags.email; }
      if (email) {
        email = email.split(";")[0];
        if (!validateEmailAddress(email)) { properties.issues.push(`E-Mail is not valid: ${email}`); }
      }
      if ("contact:email" in tags && "email" in tags) { properties.issues.push("'email' and 'contact:email' defined -> remove one"); }
      ["contact:phone", "contact:mobile", "phone"].forEach((tag) => {
        if (tag in tags) { checkPhoneNumber(properties.issues, tag, tags); }
      });
      if ("contact:phone" in tags && "phone" in tags) { properties.issues.push("'phone' and 'contact:phone' defined -> remove one"); }
      if ("mobile" in tags) { properties.issues.push("'mobile' tag is not an official tag and should therefore not be used. -> change it to 'contact:mobile'"); }
      let openingHours = "undefined";
      if ("opening_hours:covid19" in tags && !["same", "restricted"].includes(tags["opening_hours:covid19"])) { openingHours = tags["opening_hours:covid19"]; }
      else if ("opening_hours" in tags) { openingHours = tags.opening_hours; }
      else { properties.undefined.push("opening_hours"); }
      if (openingHours !== "undefined") { checkOpeningHours(openingHours, properties.issues); }
      if (containsDisused(tags)) { properties.issues.push("There is a 'disused' tag: Check whether this tag is correct. If so please remove the diet tags."); }
      if ("fixme" in tags) { properties.issues.push(`fixme: ${tags.fixme}`); }
    }
    properties.issue_count = properties.issues.length + properties.undefined.length;
    if (!properties.issues.length) { delete properties.issues; }
    if (!properties.undefined.length) { delete properties.undefined; }
    if (properties.issue_count > 0) { features.push({ type: "Feature", properties, geometry }); }
  }
  process.stdout.write(`\n${total} elements.\n`);
  return { _timestamp: TIMESTAMP, type: "FeatureCollection", features: sortByIssueCount(features) };
}

async function loadAndPruneUrlData() {
  const data = (await readJson(URL_DATA_FILE)) || {};
  const today = new Date(DATE);
  for (const url of Object.keys(data)) {
    try {
      const date = new Date(data[url].date);
      const deltaDays = (today - date) / 86400000;
      if (deltaDays > MIN_URL_CHECK_AGE) { delete data[url]; }
    }
    catch { delete data[url]; }
  }
  return data;
}

async function main() {
  proc.urlData = await loadAndPruneUrlData();
  const osmData = await readJson(OVERPASS_FILE);
  if (!osmData) {
    console.error("A problem has occurred. osm_data is None");
    process.exitCode = 1;
    return;
  }
  const checkResult = await checkData(osmData);
  await writeJson(VEGGIEPLACES_CHECK_RESULT_FILE, checkResult);
  await writeJson(URL_DATA_FILE, proc.urlData);
}

process.on("unhandledRejection", (reason) => {
  console.error("Unhandled rejection suppressed:", reason);
});

if (import.meta.url === `file://${process.argv[1]}`) {
  main().catch((err) => {
    console.error("datacheck failed:", err);
    process.exit(1);
  });
}
